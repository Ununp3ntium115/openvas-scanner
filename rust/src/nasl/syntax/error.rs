// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

//! Defines TokenError and its companion macros.

use std::{io, ops::Range};

use thiserror::Error;

use crate::nasl::{
    error::AsCodespanError,
    syntax::{token::Token, Statement},
};

use super::{tokenizer::TokenizerErrorKind, TokenizerError};

#[derive(Clone, Debug, PartialEq, Eq, Error)]
/// A list specifying general categories of Syntax error.
pub enum ErrorKind {
    /// An unexpected token occurred
    #[error("Unexpected token: {0}")]
    UnexpectedToken(Token),
    /// A token is unclosed
    ///
    /// Could happen on string literals.
    #[error("Unclosed token: {0}")]
    UnclosedToken(Token),
    /// An unexpected statement occurred
    #[error("Unexpected statement: {0}")]
    UnexpectedStatement(Statement),
    /// When an unexpected Statement occurs it maybe an MissingSemicolon
    #[error("Missing semicolon: {0}")]
    MissingSemicolon(Statement),
    /// An token is unclosed
    #[error("Unclosed statement: {0}")]
    UnclosedStatement(Statement),
    /// Maximal recursion depth reached. Simplify NASL code.
    #[error("Maximal recursion depth of {0} reached, the NASL script is too complex.")]
    MaxRecursionDepth(u8),
    /// The cursor is already at the end but that is not expected
    #[error("Unexpected end of file.")]
    EoF,
    /// An IO Error occurred while loading a NASL file
    #[error("IOError: {0}")]
    IOError(io::ErrorKind),
    #[error("{0}")]
    Tokenizer(#[from] TokenizerErrorKind),
}

impl From<TokenizerError> for SyntaxError {
    fn from(value: TokenizerError) -> Self {
        Self {
            kind: ErrorKind::Tokenizer(value.kind),
            range: value.range,
        }
    }
}

/// Is used to express errors while parsing.
#[derive(Clone, Debug, PartialEq, Eq, Error)]
#[error("{kind}")]
pub struct SyntaxError {
    /// A human readable reason why this error is returned
    pub kind: ErrorKind,
    pub range: Range<usize>,
}

// TODO Remove this
impl SyntaxError {
    /// Returns a token of the underlying error kind
    pub fn as_token(&self) -> Option<&Token> {
        match &self.kind {
            ErrorKind::UnexpectedToken(t) => Some(t),
            ErrorKind::UnclosedToken(t) => Some(t),
            ErrorKind::UnexpectedStatement(s) => Some(s.as_token()),
            ErrorKind::MissingSemicolon(s) => Some(s.as_token()),
            ErrorKind::UnclosedStatement(s) => Some(s.as_token()),
            ErrorKind::EoF => None,
            ErrorKind::IOError(_) => None,
            ErrorKind::MaxRecursionDepth(_) => None,
            ErrorKind::Tokenizer(_) => None,
        }
    }
}

#[macro_export]
macro_rules! syntax_error {
    ($kind:expr) => {{
        use $crate::nasl::syntax::SyntaxError;
        // TODO
        SyntaxError {
            kind: $kind,
            range: 0..0,
        }
    }};
}

#[macro_export]
macro_rules! unexpected_token {
    ($token:expr) => {{
        use $crate::nasl::syntax::ErrorKind;
        use $crate::syntax_error;
        syntax_error!(ErrorKind::UnexpectedToken($token))
    }};
}

#[macro_export]
macro_rules! unexpected_statement {
    ($statement:expr) => {{
        use $crate::nasl::syntax::ErrorKind;
        use $crate::syntax_error;
        syntax_error!(ErrorKind::MissingSemicolon($statement))
    }};
}

#[macro_export]
macro_rules! unclosed_statement {
    ($statement:expr) => {{
        use $crate::nasl::syntax::ErrorKind;
        use $crate::syntax_error;

        syntax_error!(ErrorKind::UnclosedStatement($statement))
    }};
}

#[macro_export]
macro_rules! unclosed_token {
    ($token:expr) => {{
        use $crate::nasl::syntax::ErrorKind;
        use $crate::syntax_error;

        syntax_error!(ErrorKind::UnclosedToken($token))
    }};
}

#[macro_export]
macro_rules! unexpected_end {
    ($reason:expr) => {{
        use $crate::nasl::syntax::ErrorKind;
        use $crate::syntax_error;
        syntax_error!(ErrorKind::EoF)
    }};
}

/// Creates an maximal recursion depth reached error.
///
/// To prevent stack overflows the Lexer veriefies it's depth and returns an error.
///
/// # Examples
///
/// Basic usage:
/// ```rust
/// use scannerlib::max_recursion;
/// max_recursion!(255);
/// ```
#[macro_export]
macro_rules! max_recursion {
    ($reason:expr) => {{
        use $crate::nasl::syntax::ErrorKind;
        use $crate::syntax_error;
        syntax_error!(ErrorKind::MaxRecursionDepth($reason))
    }};
}

impl SyntaxError {
    /// Returns the ErrorKind of SyntaxError
    pub fn kind(&self) -> &ErrorKind {
        &self.kind
    }
}

impl AsCodespanError for SyntaxError {
    fn range(&self) -> std::ops::Range<usize> {
        self.range.clone()
    }

    fn message(&self) -> String {
        format!("{}", self.kind)
    }
}

#[cfg(test)]
mod tests {
    use crate::nasl::syntax::{parse_only_first_error, ErrorKind, TokenKind};

    fn test_for_missing_semicolon(code: &str) {
        let err = parse_only_first_error(code).unwrap_err();
        match err.kind {
            ErrorKind::MissingSemicolon(_) => {}
            _ => panic!("Expected MissingSemicolon but got: {err:?}"),
        }
    }

    fn test_for_unclosed_token(code: &str, kind: TokenKind) {
        let err = parse_only_first_error(code).unwrap_err();
        match err.kind {
            ErrorKind::UnclosedToken(token) => {
                assert_eq!(token.kind(), &kind);
            }
            _ => panic!("Expected UnclosedToken but got: {err:?} for {code}"),
        }
    }
    #[test]
    fn missing_semicolon_assignment() {
        let code = "a = 12";
        test_for_missing_semicolon(code);
        let code = "a = [1, 2, 4]";
        test_for_missing_semicolon(code);
    }

    #[test]
    fn missing_semicolon_call() {
        let code = "called(me)";
        test_for_missing_semicolon(code);
    }

    #[test]
    fn missing_right_paren() {
        test_for_unclosed_token("called(me;", TokenKind::LeftParen);
        test_for_unclosed_token("foreach a(x { a = 2;", TokenKind::LeftParen);
        test_for_unclosed_token("for (i = 0; i < 10; i++ ;", TokenKind::LeftParen);
        test_for_unclosed_token("while (TRUE ;", TokenKind::LeftParen);
    }

    #[test]
    fn missing_right_curly_bracket() {
        test_for_unclosed_token("if (a) { a = 2", TokenKind::LeftCurlyBracket);
        test_for_unclosed_token("foreach a(x) { a = 2;", TokenKind::LeftCurlyBracket);
        test_for_unclosed_token("{ a = 2;", TokenKind::LeftCurlyBracket);
        test_for_unclosed_token("function a() { a = 2;", TokenKind::LeftCurlyBracket);
    }
}
