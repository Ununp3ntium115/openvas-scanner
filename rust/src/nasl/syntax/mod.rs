// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

#![doc = include_str!("README.md")]
mod cursor;
mod error;
mod grouping_extension;
mod keyword_extension;
mod lexer;
mod loader;
mod naslvalue;
mod operation;
mod prefix_extension;
mod statement;
mod token;
mod tokenizer;
mod variable_extension;

pub use crate::storage::item::ACT;
pub use error::{ErrorKind, SyntaxError};
pub use lexer::Lexer;
pub use loader::*;
pub use naslvalue::*;
pub use statement::*;
pub use token::Keyword;
pub use token::Token;
pub use token::TokenKind;
pub use tokenizer::Tokenizer;

/// Parses given code and returns found Statements and Errors
///
/// # Examples
/// Basic usage:
///
/// ```
/// use scannerlib::nasl::syntax::{Statement, SyntaxError, parse};
/// let statements =
///     parse("a = 23;b = 1;").collect::<Vec<Result<Statement, SyntaxError>>>();
/// ````
pub fn parse(code: &str) -> impl Iterator<Item = Result<Statement, SyntaxError>> + '_ {
    // TODO Do not unwrap here, handle errors properly.
    let tokens = Tokenizer::tokenize(code).unwrap();
    Lexer::new(tokens)
}

#[cfg(test)]
mod tests {
    use crate::nasl::syntax::{
        cursor::Cursor,
        token::{Keyword, Token, TokenKind},
        Tokenizer,
    };

    #[test]
    fn use_cursor() {
        let mut cursor = Cursor::new("  \n\tdisplay(12);");
        cursor.skip_while(|c| c.is_whitespace());
        assert_eq!(cursor.advance(), Some('d'));
    }

    #[test]
    fn use_tokenizer() {
        let all_tokens = Tokenizer::tokenize("local_var hello = 'World!';").unwrap();
        assert_eq!(
            all_tokens,
            vec![
                Token {
                    kind: TokenKind::Identifier(Keyword::LocalVar),
                    line_column: (1, 1),
                    position: (0, 9)
                },
                Token {
                    kind: TokenKind::Identifier(Keyword::Undefined("hello".to_owned())),
                    line_column: (1, 11),
                    position: (10, 15)
                },
                Token {
                    kind: TokenKind::Equal,
                    line_column: (1, 17),
                    position: (16, 17)
                },
                Token {
                    kind: TokenKind::Data("World!".as_bytes().to_vec()),
                    line_column: (1, 19),
                    position: (18, 26)
                },
                Token {
                    kind: TokenKind::Semicolon,
                    line_column: (1, 27),
                    position: (26, 27)
                },
            ]
        );
    }

    #[test]
    fn use_parser() {
        let code = "a = 23;b = 1;";
        let expected = ["a = 23;", "b = 1;"];
        for (i, s) in super::parse(code).enumerate() {
            let stmt = s.unwrap();
            //assert!(matches!(stmt.kind(), Assign(..)));
            assert_eq!(&code[stmt.range()], expected[i]);
        }
    }
}
