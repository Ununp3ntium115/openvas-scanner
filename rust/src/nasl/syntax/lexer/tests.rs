use crate::nasl::{
    syntax::{grammar::Declaration, Statement},
    Code,
};

pub fn parse_ok(file_name: &str, code: &str) -> Vec<Declaration> {
    let results = Code::from_string_fake_filename(code, file_name).parse();
    results.emit_errors().unwrap().decls()
}

pub fn parse_err(file_name: &str, code: &str) -> String {
    let results = Code::from_string_fake_filename(code, file_name).parse();
    results.unwrap_errors_str()
}

macro_rules! parse_test_ok {
    ($name: ident, $($code: literal$(,)?)*) => {
        #[test]
        fn $name() {
            $(
                insta::assert_snapshot!(crate::nasl::syntax::lexer::tests::parse_ok(
                    stringify!($name),
                        $code
                )
                .into_iter()
                .map(|stmt| stmt.to_string())
                .collect::<Vec<_>>()
                .join("\n"));
            )*
        }
    };
}

macro_rules! parse_test_err {
    ($name: ident, $($code: literal$(,)?)*) => {
        #[test]
        fn $name() {
            $(
                insta::assert_snapshot!(crate::nasl::syntax::lexer::tests::parse_err(
                    stringify!($name),
                        $code
                ));
            )*
        }
    };
}

pub(crate) use {parse_test_err, parse_test_ok};

use crate::nasl::syntax::token;

use super::*;

use StatementKind::*;
use TokenKind::*;

// simplified resolve method to verify a calculate with a given statement
fn resolve(s: &Statement) -> i64 {
    let callable = |stmts: &[Statement], calculus: Box<dyn Fn(i64, i64) -> i64>| -> i64 {
        let right = &stmts[1];
        let left = &stmts[0];
        calculus(resolve(left), resolve(right))
    };
    let single_callable = |stmts: &[Statement], calculus: Box<dyn Fn(i64) -> i64>| -> i64 {
        let left = &stmts[0];
        calculus(resolve(left))
    };
    match s.kind() {
        Primitive => match s.start().kind() {
            Literal(token::Literal::Number(num)) => *num,
            _ => todo!(),
        },
        Operator(head, rest) => match head {
            Tilde => single_callable(rest, Box::new(|left| !left)),
            Plus => callable(rest, Box::new(|left, right| left + right)),
            Minus if rest.len() == 1 => single_callable(rest, Box::new(|left| -left)),
            Minus => callable(rest, Box::new(|left, right| left - right)),
            Star => callable(rest, Box::new(|left, right| left * right)),
            Slash => callable(rest, Box::new(|left, right| left / right)),
            Percent => callable(rest, Box::new(|left, right| left % right)),
            LessLess => callable(rest, Box::new(|left, right| left << right)),
            GreaterGreater => callable(rest, Box::new(|left, right| left >> right)),
            Ampersand => callable(rest, Box::new(|left, right| left & right)),
            Pipe => callable(rest, Box::new(|left, right| left | right)),
            Caret => callable(rest, Box::new(|left, right| left ^ right)),
            GreaterGreaterGreater => {
                callable(
                    rest,
                    Box::new(|left, right| {
                        // this operator is used to drop signed bits
                        // so the result depends heavily if it is u32, u64, ...
                        // to have the same results as in javascript we use u32 in this example
                        let left_casted = left as u32;
                        (left_casted >> right) as i64
                    }),
                )
            }
            StarStar => callable(
                rest,
                Box::new(|left, right| (left as u32).pow(right as u32) as i64),
            ),
            token => {
                todo!("{:?}", token)
            }
        },
        _ => todo!("operator not found"),
    }
}

macro_rules! calculated_test {
    ($code:expr, $expected:expr) => {
        let expr = match parse_ok("", $code).remove(0) {
            Declaration::Statement(stmt) => stmt,
        };
        assert_eq!(resolve(&expr), $expected);
    };
}

#[test]
fn ordering() {
    calculated_test!("1 + 5 * 6;", 31);
    calculated_test!("3 * 10 + 10 / 5;", 32);
    calculated_test!("3 * 10 / 5;", 6);
    calculated_test!("3 * 10 / 5 % 4;", 2);
    calculated_test!("1 - 1 - 1;", -1);
    calculated_test!("1 - 1 * 2;", -1);
}

#[test]
fn grouping() {
    calculated_test!("(2 + 5) * 2;", 14);
}

#[test]
fn pow() {
    calculated_test!("2 ** 4;", 16);
}

#[test]
fn bitwise_operations() {
    //shifting
    calculated_test!("1 << 2 * 3;", 64);
    calculated_test!("3 * 12 >> 2;", 9);
    calculated_test!("-5 >>> 2;", 1073741822);
    // operations
    calculated_test!("1 & 0;", 0);
    calculated_test!("~1 | 0;", -2);
    calculated_test!("1 ^ 1;", 0);
}

parse_test_ok!(
    operator_assignment,
    "
    a += 1;
    a -= 1;
    a /= 1;
    a *= 1;
    a %= 1;
    a >>= 1;
    a <<= 1;
    a >>>= 1;
    "
);

parse_test_ok!(
    compare_operator,
    "
    a !~ '1';
    a =~ '1';
    a >< '1';
    a >!< '1';
    a == '1';
    a != '1';
    a > '1';
    a < '1';
    a >= '1';
    a <= '1';
    x() x 2;
    "
);

parse_test_ok!(logical_operator, "a && 1; a || 1;");

parse_test_ok!(assignment, "(a = 1);");

parse_test_ok!(variable_assignment_operator, "a++; a--; a[1]++; a[1]--;");

parse_test_ok!(primitive, "1;");
parse_test_ok!(variable, "a;");
parse_test_ok!(array, "a[1];");
parse_test_ok!(call, "a();");
parse_test_ok!(exit, "exit(0);");
parse_test_ok!(return_stmt, "return 0;");
parse_test_ok!(break_stmt, "break;");
parse_test_ok!(continue_stmt, "continue;");
parse_test_ok!(include, "include(\"test.inc\");");
parse_test_ok!(declare, "local_var a;");
parse_test_ok!(parameter, "[a, b];");
parse_test_ok!(named_parameter, "a: b;");
parse_test_ok!(assign, "a = 1;");
parse_test_ok!(add, "a + 1;");
parse_test_ok!(sub, "a - 1;");
parse_test_ok!(mul, "a * 1;");
parse_test_ok!(div, "a / 1;");
parse_test_ok!(modulo, "a % 1;");
parse_test_ok!(return_assign, "a++;");
parse_test_ok!(assign_return, "--a;");
parse_test_ok!(if_stmt, "if (a) b; else c;");
parse_test_ok!(for_stmt, "for (i = 0; i < 10; i++) a;");
parse_test_ok!(while_stmt, "while (a) b;");
parse_test_ok!(repeat, "repeat a; until b;");
parse_test_ok!(foreach, "foreach a(b) c;");
parse_test_ok!(block, "{ a; }");
parse_test_ok!(function_declaration, "function a(b) {c;}");
parse_test_ok!(no_op, ";");

parse_test_ok!(variables, "a;");
parse_test_ok!(arrays, "a[0];", "a = [1, 2, 3];", "a[0] = [1, 2, 4];");
parse_test_ok!(anon_function_call, "a(1, 2, 3);");
parse_test_ok!(
    named_function_call,
    "script_tag(name:\"cvss_base\", value:1 + 1 % 2);"
);
parse_test_ok!(use_parser, "a = 23;b = 1;");

// test_err!(wrong_assignment, "a = ");
// test_err!(wrong_keyword_assignment, "a = for;");

parse_test_ok!(
    position,
    r#"
        a = 1 + 1;
        b = 2 * 2;
        a = ++a;
        arr = mkarray(a, b, c      );
        arr[++a];
        exit(1);
        return 1;
        include('test.inc');
        local_var a, b, c;
        global_var a, b, c;
        if (a) display(1); else display(2);
        for (i = 1; i < 10; i++) display(i);
        while(TRUE) display(i);
        foreach a(q) display(a);
        repeat display("q"); until 1;
        {
            a;
            b;
        }
        function register_packages( buf ) { return 1; }
    "#
);

parse_test_ok!(
    change_to_peek,
    "send_packet( udp, pcap_active:FALSE ) x 200;"
);

parse_test_ok!(
    stack_overflow,
    r###"
        req = raw_string(0x00, 0x00, 0x03, 0x14, 0x08, 0x14, 0xff, 0x9f,
                0xde, 0x5d, 0x5f, 0xb3, 0x07, 0x8f, 0x49, 0xa7,
                0x79, 0x6a, 0x03, 0x3d, 0xaf, 0x55, 0x00, 0x00,
                0x00, 0x7e, 0x64, 0x69, 0x66, 0x66, 0x69, 0x65,
                0x2d, 0x68, 0x65, 0x6c, 0x6c, 0x6d, 0x61, 0x6e,
                0x2d, 0x67, 0x72, 0x6f, 0x75, 0x70, 0x2d, 0x65,
                0x78, 0x63, 0x68, 0x61, 0x6e, 0x67, 0x65, 0x2d,
                0x73, 0x68, 0x61, 0x32, 0x35, 0x36, 0x2c, 0x64,
                0x69, 0x66, 0x66, 0x69, 0x65, 0x2d, 0x68, 0x65,
                0x6c, 0x6c, 0x6d, 0x61, 0x6e, 0x2d, 0x67, 0x72,
                0x6f, 0x75, 0x70, 0x2d, 0x65, 0x78, 0x63, 0x68,
                0x61, 0x6e, 0x67, 0x65, 0x2d, 0x73, 0x68, 0x61,
                0x31, 0x2c, 0x64, 0x69, 0x66, 0x66, 0x69, 0x65,
                0x2d, 0x68, 0x65, 0x6c, 0x6c, 0x6d, 0x61, 0x6e,
                0x2d, 0x67, 0x72, 0x6f, 0x75, 0x70, 0x31, 0x34,
                0x2d, 0x73, 0x68, 0x61, 0x31, 0x2c, 0x64, 0x69,
                0x66, 0x66, 0x69, 0x65, 0x2d, 0x68, 0x65, 0x6c,
                0x6c, 0x6d, 0x61, 0x6e, 0x2d, 0x67, 0x72, 0x6f,
                0x75, 0x70, 0x31, 0x2d, 0x73, 0x68, 0x61, 0x31,
                0x00, 0x00, 0x00, 0x0f, 0x73, 0x73, 0x68, 0x2d,
                0x72, 0x73, 0x61, 0x2c, 0x73, 0x73, 0x68, 0x2d,
                0x64, 0x73, 0x73, 0x00, 0x00, 0x00, 0x9d, 0x61,
                0x65, 0x73, 0x31, 0x32, 0x38, 0x2d, 0x63, 0x62,
                0x63, 0x2c, 0x33, 0x64, 0x65, 0x73, 0x2d, 0x63,
                0x62, 0x63, 0x2c, 0x62, 0x6c, 0x6f, 0x77, 0x66,
                0x69, 0x73, 0x68, 0x2d, 0x63, 0x62, 0x63, 0x2c,
                0x63, 0x61, 0x73, 0x74, 0x31, 0x32, 0x38, 0x2d,
                0x63, 0x62, 0x63, 0x2c, 0x61, 0x72, 0x63, 0x66,
                0x6f, 0x75, 0x72, 0x31, 0x32, 0x38, 0x2c, 0x61,
                0x72, 0x63, 0x66, 0x6f, 0x75, 0x72, 0x32, 0x35,
                0x36, 0x2c, 0x61, 0x72, 0x63, 0x66, 0x6f, 0x75,
                0x72, 0x2c, 0x61, 0x65, 0x73, 0x31, 0x39, 0x32,
                0x2d, 0x63, 0x62, 0x63, 0x2c, 0x61, 0x65, 0x73,
                0x32, 0x35, 0x36, 0x2d, 0x63, 0x62, 0x63, 0x2c,
                0x72, 0x69, 0x6a, 0x6e, 0x64, 0x61, 0x65, 0x6c,
                0x2d, 0x63, 0x62, 0x63, 0x40, 0x6c, 0x79, 0x73,
                0x61, 0x74, 0x6f, 0x72, 0x2e, 0x6c, 0x69, 0x75,
                0x2e, 0x73, 0x65, 0x2c, 0x61, 0x65, 0x73, 0x31,
                0x32, 0x38, 0x2d, 0x63, 0x74, 0x72, 0x2c, 0x61,
                0x65, 0x73, 0x31, 0x39, 0x32, 0x2d, 0x63, 0x74,
                0x72, 0x2c, 0x61, 0x65, 0x73, 0x32, 0x35, 0x36,
                0x2d, 0x63, 0x74, 0x72, 0x00, 0x00, 0x00, 0x9d,
                0x61, 0x65, 0x73, 0x31, 0x32, 0x38, 0x2d, 0x63,
                0x62, 0x63, 0x2c, 0x33, 0x64, 0x65, 0x73, 0x2d,
                0x63, 0x62, 0x63, 0x2c, 0x62, 0x6c, 0x6f, 0x77,
                0x66, 0x69, 0x73, 0x68, 0x2d, 0x63, 0x62, 0x63,
                0x2c, 0x63, 0x61, 0x73, 0x74, 0x31, 0x32, 0x38,
                0x2d, 0x63, 0x62, 0x63, 0x2c, 0x61, 0x72, 0x63,
                0x66, 0x6f, 0x75, 0x72, 0x31, 0x32, 0x38, 0x2c,
                0x61, 0x72, 0x63, 0x66, 0x6f, 0x75, 0x72, 0x32,
                0x35, 0x36, 0x2c, 0x61, 0x72, 0x63, 0x66, 0x6f,
                0x75, 0x72, 0x2c, 0x61, 0x65, 0x73, 0x31, 0x39,
                0x32, 0x2d, 0x63, 0x62, 0x63, 0x2c, 0x61, 0x65,
                0x73, 0x32, 0x35, 0x36, 0x2d, 0x63, 0x62, 0x63,
                0x2c, 0x72, 0x69, 0x6a, 0x6e, 0x64, 0x61, 0x65,
                0x6c, 0x2d, 0x63, 0x62, 0x63, 0x40, 0x6c, 0x79,
                0x73, 0x61, 0x74, 0x6f, 0x72, 0x2e, 0x6c, 0x69,
                0x75, 0x2e, 0x73, 0x65, 0x2c, 0x61, 0x65, 0x73,
                0x31, 0x32, 0x38, 0x2d, 0x63, 0x74, 0x72, 0x2c,
                0x61, 0x65, 0x73, 0x31, 0x39, 0x32, 0x2d, 0x63,
                0x74, 0x72, 0x2c, 0x61, 0x65, 0x73, 0x32, 0x35,
                0x36, 0x2d, 0x63, 0x74, 0x72, 0x00, 0x00, 0x00,
                0x69, 0x68, 0x6d, 0x61, 0x63, 0x2d, 0x6d, 0x64,
                0x35, 0x2c, 0x68, 0x6d, 0x61, 0x63, 0x2d, 0x73,
                0x68, 0x61, 0x31, 0x2c, 0x75, 0x6d, 0x61, 0x63,
                0x2d, 0x36, 0x34, 0x40, 0x6f, 0x70, 0x65, 0x6e,
                0x73, 0x73, 0x68, 0x2e, 0x63, 0x6f, 0x6d, 0x2c,
                0x68, 0x6d, 0x61, 0x63, 0x2d, 0x72, 0x69, 0x70,
                0x65, 0x6d, 0x64, 0x31, 0x36, 0x30, 0x2c, 0x68,
                0x6d, 0x61, 0x63, 0x2d, 0x72, 0x69, 0x70, 0x65,
                0x6d, 0x64, 0x31, 0x36, 0x30, 0x40, 0x6f, 0x70,
                0x65, 0x6e, 0x73, 0x73, 0x68, 0x2e, 0x63, 0x6f,
                0x6d, 0x2c, 0x68, 0x6d, 0x61, 0x63, 0x2d, 0x73,
                0x68, 0x61, 0x31, 0x2d, 0x39, 0x36, 0x2c, 0x68,
                0x6d, 0x61, 0x63, 0x2d, 0x6d, 0x64, 0x35, 0x2d,
                0x39, 0x36, 0x00, 0x00, 0x00, 0x69, 0x68, 0x6d,
                0x61, 0x63, 0x2d, 0x6d, 0x64, 0x35, 0x2c, 0x68,
                0x6d, 0x61, 0x63, 0x2d, 0x73, 0x68, 0x61, 0x31,
                0x2c, 0x75, 0x6d, 0x61, 0x63, 0x2d, 0x36, 0x34,
                0x40, 0x6f, 0x70, 0x65, 0x6e, 0x73, 0x73, 0x68,
                0x2e, 0x63, 0x6f, 0x6d, 0x2c, 0x68, 0x6d, 0x61,
                0x63, 0x2d, 0x72, 0x69, 0x70, 0x65, 0x6d, 0x64,
                0x31, 0x36, 0x30, 0x2c, 0x68, 0x6d, 0x61, 0x63,
                0x2d, 0x72, 0x69, 0x70, 0x65, 0x6d, 0x64, 0x31,
                0x36, 0x30, 0x40, 0x6f, 0x70, 0x65, 0x6e, 0x73,
                0x73, 0x68, 0x2e, 0x63, 0x6f, 0x6d, 0x2c, 0x68,
                0x6d, 0x61, 0x63, 0x2d, 0x73, 0x68, 0x61, 0x31,
                0x2d, 0x39, 0x36, 0x2c, 0x68, 0x6d, 0x61, 0x63,
                0x2d, 0x6d, 0x64, 0x35, 0x2d, 0x39, 0x36, 0x00,

                ##3rd byte in this next line causes crash
                0x00, 0x00, 0x28, 0x7a, 0x6c, 0x69, 0x62, 0x40,
                0x6f, 0x70, 0x65, 0x6e, 0x73, 0x73, 0x68, 0x2e,
                0x63, 0x6f, 0x6d, 0x2c, 0x7a, 0x6c, 0x69, 0x62,
                0x2c, 0x6e, 0x6f, 0x6e, 0x65, 0x00, 0x00, 0x00,
                0x1a, 0x7a, 0x6c, 0x69, 0x62, 0x40, 0x6f, 0x70,
                0x65, 0x6e, 0x73, 0x73, 0x68, 0x2e, 0x63, 0x6f,
                0x6d, 0x2c, 0x7a, 0x6c, 0x69, 0x62, 0x2c, 0x6e,
                0x6f, 0x6e, 0x65, 0x00, 0x00, 0x00, 0x00, 0x00,
                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
                0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
                0x0a);
    "###
);

parse_test_ok!(
    jsp_example,
    r#"
    gms_path = gms_path + 'webapps\\appliance\\';
    jsp = '<% out.println( "' + jsp_print  + '" ); %>';
    "#
);

parse_test_ok!(
    unexpected_assign,
    r###"
    while(y = recv(socket:soc, length:1024)) {
    buf1 += y;
    }
    "###
);

parse_test_ok!(
    unexpected_noop,
    r###"
    if( ! version || version == '' ) return;
    "###
);

parse_test_ok!(
    unexpected_equal_operator,
    r###"
# Message Server runs on ports 36xx or 39xx
if (port < 3600 || port >= 3700)
if (port < 3900 || port >= 4000)
exit(0);

soc = open_sock_tcp(port);

    "###
);

parse_test_ok!(
    unexpected_plusplus,
    r###"
    cookie_jar[this_cookie]++;
    "###
);
