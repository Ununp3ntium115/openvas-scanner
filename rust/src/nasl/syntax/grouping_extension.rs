// SPDX-FileCopyrightText: 2023 Greenbone AG
//
// SPDX-License-Identifier: GPL-2.0-or-later WITH x11vnc-openssl-exception

use super::{
    AssignOrder, Statement, StatementKind,
    error::SyntaxError,
    lexer::{End, Lexer},
    token::{Token, TokenKind},
};

use crate::{unclosed_token, unexpected_token};

pub(crate) trait Grouping {
    /// Parses (...)
    fn parse_paren(&mut self, token: Token) -> Result<Statement, SyntaxError>;
    /// Parses {...}
    fn parse_block(&mut self, token: Token) -> Result<Statement, SyntaxError>;
    /// General Grouping parsing. Is called within prefix_extension.
    fn parse_grouping(&mut self, token: Token) -> Result<(End, Statement), SyntaxError>;
}

impl Grouping for Lexer {
    fn parse_paren(&mut self, token: Token) -> Result<Statement, SyntaxError> {
        let (end, right) = self.statement(0, &|cat| cat == &TokenKind::RightParen)?;

        match end {
            End::Done(end) => {
                self.depth = 0;
                Ok(match right.kind() {
                    StatementKind::Assign(cat, _, first, second) => {
                        Statement::with_start_end_token(
                            token,
                            end,
                            StatementKind::Assign(
                                cat.clone(),
                                AssignOrder::AssignReturn,
                                first.clone(),
                                second.clone(),
                            ),
                        )
                    }

                    _ => right,
                })
            }

            End::Continue => Err(unclosed_token!(token)),
        }
    }

    fn parse_block(&mut self, kw: Token) -> Result<Statement, SyntaxError> {
        let mut results = vec![];
        while let Some(token) = self.peek() {
            if token.kind() == &TokenKind::RightCurlyBracket {
                let _ = self.token();

                self.depth = 0;
                let stmt = Statement::with_start_end_token(
                    kw,
                    token.clone(),
                    StatementKind::Block(results),
                );
                return Ok(stmt);
            }
            let (end, stmt) = self.statement(0, &|cat| cat == &TokenKind::Semicolon)?;
            if end.is_done() && !matches!(stmt.kind(), StatementKind::NoOp) {
                results.push(stmt);
            }
        }
        Err(unclosed_token!(kw))
    }

    fn parse_grouping(&mut self, token: Token) -> Result<(End, Statement), SyntaxError> {
        fn as_done(stmt: Statement) -> (End, Statement) {
            (End::Done(stmt.end().clone()), stmt)
        }

        fn as_con(stmt: Statement) -> (End, Statement) {
            (End::Continue, stmt)
        }
        match token.kind() {
            TokenKind::LeftParen => self.parse_paren(token).map(as_con),
            TokenKind::LeftCurlyBracket => self.parse_block(token).map(as_done),
            TokenKind::LeftBrace => {
                let (end, right) = self.parse_comma_group(TokenKind::RightBrace)?;
                match end {
                    End::Done(end) => Ok((
                        End::Continue,
                        Statement::with_start_end_token(
                            token,
                            end,
                            StatementKind::Parameter(right),
                        ),
                    )),
                    End::Continue => Err(unclosed_token!(token)),
                }
            }
            _ => Err(unexpected_token!(token)),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::nasl::syntax::lexer::tests::parse_test_ok;

    parse_test_ok!(
        variables,
        r"
            {
                a = b + 1;
                b = a - --c;
                {
                   d = 23;
                }
            }
            "
    );
}
