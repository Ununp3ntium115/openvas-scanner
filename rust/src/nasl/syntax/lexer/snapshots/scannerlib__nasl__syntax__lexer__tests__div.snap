---
source: src/nasl/syntax/lexer/tests.rs
expression: "parse_ok(stringify! (div), \"a / 1;\")"
snapshot_kind: text
---
[
    Statement {
        start: Token {
            kind: Ident(
                Ident(
                    "a",
                ),
            ),
            position: (
                0,
                1,
            ),
        },
        end: Some(
            Token {
                kind: Semicolon,
                position: (
                    5,
                    6,
                ),
            },
        ),
        kind: Operator(
            Slash,
            [
                Statement {
                    start: Token {
                        kind: Ident(
                            Ident(
                                "a",
                            ),
                        ),
                        position: (
                            0,
                            1,
                        ),
                    },
                    end: Some(
                        Token {
                            kind: Ident(
                                Ident(
                                    "a",
                                ),
                            ),
                            position: (
                                0,
                                1,
                            ),
                        },
                    ),
                    kind: Variable,
                },
                Statement {
                    start: Token {
                        kind: Literal(
                            Number(
                                1,
                            ),
                        ),
                        position: (
                            4,
                            5,
                        ),
                    },
                    end: Some(
                        Token {
                            kind: Semicolon,
                            position: (
                                5,
                                6,
                            ),
                        },
                    ),
                    kind: Primitive,
                },
            ],
        ),
    },
]
